src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -109,5 +117,120 @@ public Solution GetCompileTimeSolution(Solution designTimeSolution)
                 return _lazyCompileTimeSolution;
             }
         }
+
+        // Copied from
+        // https://github.com/dotnet/sdk/blob/main/src/RazorSdk/SourceGenerators/RazorSourceGenerator.Helpers.cs#L32
+        private static string GetIdentifierFromPath(string filePath)
+        {
+            var builder = new StringBuilder(filePath.Length);
+
+            for (var i = 0; i < filePath.Length; i++)
+            {
+                switch (filePath[i])
+                {
+                    case ':' or '\\' or '/':
+                    case char ch when !char.IsLetterOrDigit(ch):
+                        builder.Append('_');
+                        break;
+                    default:
+                        builder.Append(filePath[i]);
+                        break;
+                }
+            }
+
+            return builder.ToString();
+        }
+
+        private static bool IsRazorDesignTimeDocument(DocumentState documentState)
+            => documentState.Attributes.DesignTimeOnly && documentState.FilePath?.EndsWith(".razor.g.cs") == true;
```

Comment:
(question) Is the DesignTimeOnly flag set for non-razor documents, or are there other scenarios that also set the flag?

Reply:
May be set in other cases. That's why the extension check.